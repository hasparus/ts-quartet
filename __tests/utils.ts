import { getDescription } from "./getDescription";

export function snapshot(validator: any) {
  const description = getDescription(validator);

  expect(description).toMatchSnapshot();
}
export function getExplanation(
  validator: { (arg0: any): void; explanations: any },
  value: any
) {
  validator(value);
  return validator.explanations;
}
export function isValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (!valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(true);
}
export function isNotValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(false);
}

export function tables(validator: any, valids: any[], invalids: any[]) {
  for (const valid of valids) {
    isValid(validator, valid);
  }
  for (const invalid of invalids) {
    isNotValid(validator, invalid);
  }
}
