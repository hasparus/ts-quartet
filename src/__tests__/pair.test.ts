import { v, e, quartet } from '../index'
import { puretables, snapshot, tables } from './utils'

describe('v.pair', () => {
  test('Pure v.arrayOf', () => {
    const squareNumbers = v(
      v.arrayOf(
        v.pair(
          v.and(
            {
              value: v.number,
              key: v.number,
            },
            v.custom(
              ({ key, value }: { key: number; value: any }) => value === key * key,
            ),
          ),
        ),
      ),
    )

    snapshot(squareNumbers)

    puretables(squareNumbers, [[], [0], [0, 1], [0, 1, 4]], [[1], [0, 2], [0, 1, 3]])
  })
  test('Pure v.arrayOf in complex', () => {
    const squareNumbers = v(
      v.arrayOf(
        v.and(
          v.number,
          v.pair(
            v.and(
              {
                value: v.number,
                key: v.number,
              },
              v.custom(
                ({ key, value }: { key: number; value: any }) => value === key * key,
              ),
            ),
          ),
        ),
      ),
    )

    snapshot(squareNumbers)

    puretables(squareNumbers, [[], [0], [0, 1], [0, 1, 4]], [[1], [0, 2], [0, 1, 3]])
  })
  test('Impure v.arrayOf', () => {
    const squareNumbers = v(
      v.arrayOf(
        v.pair(
          v.and(
            {
              value: v.number,
              key: v.number,
            },
            v.custom(
              ({ key, value }: { key: number; value: any }) => value === key * key,
              (value: any) => value,
            ),
          ),
        ),
      ),
    )

    snapshot(squareNumbers)

    tables(
      squareNumbers,
      [[], [0], [0, 1], [0, 1, 4]],
      [
        [[1], [{ key: 0, value: 1 }]],
        [[0, 2], [{ key: 1, value: 2 }]],
        [[0, 1, 3], [{ key: 2, value: 3 }]],
      ],
    )
  })
  test('Pure v.rest', () => {
    const validGender = ['male', 'female']
    const checkGenderDictionary = v({
      [v.rest]: v.pair({ key: validGender, value: v.number }),
    })

    snapshot(checkGenderDictionary)

    puretables(
      checkGenderDictionary,
      [{}, { male: 170 }, { female: 160 }, { male: 170, female: 160 }],
      [{ a: 123 }, { male: 170, female: 160, another: 32 }],
    )
  })
  test('Pure v.rest with another props', () => {
    const validGender = ['male', 'female']
    const checkGenderDictionary = v({
      id: v.string,
      [v.rest]: v.pair({ key: validGender, value: v.number }),
    })

    snapshot(checkGenderDictionary)

    puretables(
      checkGenderDictionary,
      [
        { id: 'id' },
        { id: 'id', male: 170 },
        { id: 'id', female: 160 },
        { id: 'id', male: 170, female: 160 },
      ],
      [
        { a: 123 },
        { male: 170 },
        { id: 1, male: 170 },
        { male: 170, female: 160, another: 32 },
      ],
    )
  })
  test('Impure v.rest', () => {
    const validGender = ['male', 'female']
    const checkGenderDictionary = v({
      [v.rest]: v.pair(
        v.custom(
          ({ key, value }: { key: any; value: any }) => {
            return validGender.includes(key) && typeof value === 'number'
          },
          (value: any) => value,
        ),
      ),
    })

    snapshot(checkGenderDictionary)

    tables(
      checkGenderDictionary,
      [{}, { male: 170 }, { female: 160 }, { male: 170, female: 160 }],
      [
        [{ a: 123 }, [{ key: 'a', value: 123 }]],
        [{ male: 170, female: 160, another: 32 }, [{ key: 'another', value: 32 }]],
      ],
    )
  })
  test('Impure v.rest with another props', () => {
    const validGender = ['male', 'female']
    const checkGenderDictionary = v({
      id: v.string,
      [v.rest]: v.pair(
        v.custom(
          ({ key, value }: { key: any; value: any }) => {
            return validGender.includes(key) && typeof value === 'number'
          },
          (value: any) => value,
        ),
      ),
    })

    snapshot(checkGenderDictionary)

    tables(
      checkGenderDictionary,
      [
        { id: 'id' },
        { id: 'id', male: 170 },
        { id: 'id', female: 160 },
        { id: 'id', male: 170, female: 160 },
      ],
      [
        [{ a: 123 }, []],
        [{ id: '123', a: 123 }, [{ key: 'a', value: 123 }]],
        [{ male: 170 }, []],
        [{ id: 1, male: 170 }, []],
        [
          { id: '123', male: 170, female: 160, another: 32 },
          [{ key: 'another', value: 32 }],
        ],
      ],
    )
  })
  test('wrong usage', () => {
    expect(() => v(v.pair({}))).toThrowErrorMatchingInlineSnapshot(
      `"Wrong usage of v.pair"`,
    )
    expect(() => v({ a: v.pair({}) })).toThrowErrorMatchingInlineSnapshot(
      `"Wrong usage of v.pair"`,
    )
    expect(() => v([v.number, v.pair({})])).toThrowErrorMatchingInlineSnapshot(
      `"Wrong usage of v.pair"`,
    )
  })
})
