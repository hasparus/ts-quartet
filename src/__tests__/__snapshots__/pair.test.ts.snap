// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v.pair Impure v.rest with another props 1`] = `
Object {
  "_": "function validator(value) {
  validator.explanations = []
  if (value == null) return false
  if (typeof value.id !== 'string') return false
  validator.keys = Object.keys(value)
  for (validator.i = 0; validator.i < validator.keys.length; validator.i++) {
    validator.key = validator.keys[validator.i]
    if (validator.omitkeys[validator.key] === true) continue
    validator.elem = value[validator.key]
    validator.keyvalue.key = validator.key;
    validator.keyvalue.value = validator.elem;
    if (!validator[\\"validator.elem\\"](validator.keyvalue)) {
      validator.explanations.push(...validator[\\"validator.elem\\"].explanations);
      return false
    }
  }
  return true
}",
  "elem": undefined,
  "explanations": Array [],
  "i": 0,
  "key": undefined,
  "keys": Array [],
  "keyvalue": Object {
    "key": undefined,
    "value": undefined,
  },
  "omitkeys": Object {
    "id": true,
  },
  "pure": false,
  "validator.elem": "function validator(value) {
  validator.explanations = []
  if (validator.custom(value)) {
    return true
  }
  validator.explvalue = validator.explanation(value,validator.custom.explanations)
  if (validator.explvalue !== undefined) {
    validator.explanations.push(validator.explvalue)
  }
  return false
}",
  "validator.elem.custom": "function (_a) {
                var key = _a.key, value = _a.value;
                return validGender.includes(key) && typeof value === 'number';
            }",
  "validator.elem.explanation": "function (value) { return value; }",
  "validator.elem.explanations": Array [],
  "validator.elem.explvalue": undefined,
  "validator.elem.pure": false,
}
`;

exports[`v.pair Pure v.arrayOf in complex 1`] = `
Object {
  "_": "function validator(value) {
  return validator.arr(value)
}",
  "arr": "function validator(value) {
  if (!value || !Array.isArray(value)) return false
  for (validator.i = 0; validator.i < value.length; validator.i++) {
    validator.elem = value[validator.i]
    if (typeof validator.elem !== 'number') return false
    validator.keyvalue.key = validator.i;
    validator.keyvalue.value = validator.elem;
    if (!validator[\\"validator.elem\\"](validator.keyvalue)) return false;
  }
  return true
}",
  "arr.elem": undefined,
  "arr.explanations": Array [],
  "arr.i": 0,
  "arr.keyvalue": Object {
    "key": undefined,
    "value": undefined,
  },
  "arr.pure": true,
  "arr.validator.elem": "function validator(value) {
  if (value == null) return false
  if (typeof value.value !== 'number') return false
  if (typeof value.key !== 'number') return false
  if (!validator.custom(value)) return false
  return true
}",
  "arr.validator.elem.custom": "function (_a) {
            var key = _a.key, value = _a.value;
            return value === key * key;
        }",
  "arr.validator.elem.explanations": Array [],
  "arr.validator.elem.pure": true,
  "explanations": Array [],
  "pure": true,
}
`;
